### PR TITLE
adding best practices for telemetry events design

### PR DESCRIPTION
**Pull request by SuYoungHong**
Added cookbook for telemetry events design best practices. RE [this bug](https://bugzilla.mozilla.org/show_bug.cgi?id=1490945)

Not sure where in the docs this should live, so put it into cookbooks by default.

### CODE REVIEW

**Comment by SuYoungHong** [comment]
@chutten , good points. I've updated the cookbook to provide more background and give some more concrete examples to make the need more clear. Let me know what you think.

File: src/cookbooks/events_best_practices.md [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+* timestamp 
+* category 
+* method
+* object 
+* value 
+* extra 
+
+With the following restrictions and features: 
+* The category, method, and object properties of any record produced by an event must have a value. 
+* All combinations values from the category, method, and object properties must be unique to that particular event (no other event can produce events with the same combination).
```
Comment (chutten) [inline src/cookbooks/events_best_practices.md]:
```suggestion
* All combinations of values from the category, method, and object properties must be unique to that particular event (no other event can be registered with the same combination). 
```

File: src/cookbooks/events_best_practices.md [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+* object 
+* value 
+* extra 
+
+With the following restrictions and features: 
+* The category, method, and object properties of any record produced by an event must have a value. 
+* All combinations values from the category, method, and object properties must be unique to that particular event (no other event can produce events with the same combination). 
+* Events can be 'turned on' or 'turned off' by it's category value. i.e. we can instruct the browser to "stop sending us events from 'devtools' category."
+
+These records are then stored in event pings and available in the events dataset.
```
Comment (chutten) [inline src/cookbooks/events_best_practices.md]: I recommend linking out to the event pipeline and dataset docs here.

File: src/cookbooks/events_best_practices.md [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+
+Another factor that can add to confusion is the fact that other events can share similar values for methods or objects (or even the combination of method and object). For example:
+
+|timestamp| category | method  | object  | value | extra | 
+|-| -------- |---------| --------|-------|-------|
+|...|normandy|update|preference_rollout|        |...|
+
+which can further confuser users. 
+
+[1]: If a user defines an event in Events.yaml without specifying a list of acceptable methods, the method will default to the name of the event for records created by that event.
```
Comment (chutten) [inline src/cookbooks/events_best_practices.md]: You may need to preface this footnote with "events do have `name` fields, but they aren't included in the event records and thus are not present in the resulting dataset. Also, If a user defines..."

File: src/cookbooks/events_best_practices.md [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+
+#### Suggested Convention: 
+
+To simplify things in the future, we suggest recording the event name in the category field in this format when designing new events: 
+
+```
+"category.event_name"
+```
+
+For example:
```
Comment (chutten) [inline src/cookbooks/events_best_practices.md]: Could you expand the example? Maybe take the opportunity to show how the `addonsManager.manage` example definition above would be changed by this convention.

File: src/cookbooks/events_best_practices.md [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+```
+
+For example: 
+* ```"navigation.search"```
+* ```"addonsManager.manage"```
+* ```"frame.tab"```
+
+
+This provides 3 advantages: 
+1. Records produced by this event will be easily identifiable. Conversely, the event which produced the record will be easier to locate in the code.
```
Comment (chutten) [inline src/cookbooks/events_best_practices.md]: "Conversely" is usually reserved for statements that oppose the previous one. I'd go with "Also" here.

File: src/cookbooks/events_best_practices.md [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+
+For example: 
+* ```"navigation.search"```
+* ```"addonsManager.manage"```
+* ```"frame.tab"```
+
+
+This provides 3 advantages: 
+1. Records produced by this event will be easily identifiable. Conversely, the event which produced the record will be easier to locate in the code. 
+2. Events can be controlled easier. The category field is what we use to "turn on" and "turn off" events. By creating a 1 to 1 mapping between categories and events, we can control events on an individual level.
```
Comment (chutten) [inline src/cookbooks/events_best_practices.md]:
```suggestion
2. Events can be controlled more easily. The category field is what we use to "turn on" and "turn off" events. By creating a 1 to 1 mapping between categories and events, we can control events on an individual level. 
```

**Comment by georgf** [comment]
@SuYoungHong @chutten 
Hey, this looks great. Could i do anything to help unblocking this from merging?

Future improvements aside, are there any blocking changes needed here? Could the rest be moved to follow-ups?

**Comment by SuYoungHong** [comment]
hey @georgf , @chutten , apologies for the long delay. 
I've made some of the suggested changes to the document. I think this is ready to merge, and we can resolve the following issue: 

* clarifying example for suggested convention using `addonsManager.manage`

In the future with a separate PR

**Comment by SuYoungHong** [comment]
looks like it's also running into issues with travis. 

It's failing tests based on spelling for the following words (filenames and event field values): 

* Events.yaml
* devtools
* total_uri_count
* addonsManager
* sideload_prompt
* normandy
* preference_rollout

how do you advise I resolve these? should I treat event field values as code in `blocks`? also for filenames?

**Comment by chutten** [comment]
Code blocks might be the most sensible way. The alternative is to add it to the list of exceptions in `.spelling`

**Comment by ethompso28** [comment]
@SuYoungHong : I got a fantastic tip from @tdsmith to install the mdspell, so that you can run it locally before submitting the PR and fix all the spelling errors before it goes through CI:

`npm install markdown-spellcheck markdown-link-check -g` 
then 
`mdspell 'src/**/*.md' --ignore-numbers --en-us --report`

**Comment by SuYoungHong** [comment]
thanks @ethompso28 ! I'll check it out!

**Comment by SuYoungHong** [comment]
cool, tests pass now. PR ready to merge if @chutten  is okay with pushing the final change request to another PR (if not, let me know and I can make the change in this PR) 

- Su

**Comment by chutten** [comment]
I'm fine with it happening as a follow-up. Thanks for your persistence in this!

**Pull request merged**